src/unit.rs

```diff
@@ -2403,21 +2403,144 @@ fn parse_type_unit_header<Endian>(input: EndianBuf<Endian>,
+        {
+            unit.offset = DebugInfoOffset(self.size() as usize);
+            self.unit(&mut unit.header, &[])
+        }
+
+        fn type_unit<'input, E>(self, unit: &mut TypeUnitHeader<'input, E>) -> Self
+            where E: Endianity
+        {
+            unit.offset = DebugTypesOffset(self.size() as usize);
+            let section = Section::with_endian(Endian::Little)
+                .L64(unit.type_signature.0)
+                .offset(unit.type_offset.0, unit.header.format);
+            let extra_header = section.get_contents().unwrap();
+            self.unit(&mut unit.header, &extra_header)
+        }
+
+        fn unit<'input, E>(self, unit: &mut UnitHeader<'input, E>, extra_header: &[u8]) -> Self
+            where E: Endianity
+        {
+            let length = Label::new();
+            let start = Label::new();
+            let end = Label::new();
+
+            let section = match unit.format {
+                Format::Dwarf32 => self.L32(&length),
+                Format::Dwarf64 => {
+                    self.L32(0xffffffff)
+                        .L64(&length)
+                }
+            };
+
+            let section = section.mark(&start)
+                .L16(unit.version)
+                .offset(unit.debug_abbrev_offset.0, unit.format)
+                .D8(unit.address_size)
+                .append_bytes(extra_header)
+                .append_bytes(unit.entries_buf.into())
+                .mark(&end);
+
+            unit.unit_length = (&end - &start) as u64;
+            length.set_const(unit.unit_length);
+
+            section
+        }
+
+        fn die<F>(self, code: u64, attr: F) -> Self
+            where F: Fn(Section) -> Section
+        {
+            let section = self.uleb(code);
+            attr(section)
+        }
+
+        fn die_null(self) -> Self {
+            self.D8(0)
+        }
+
+        fn attr_string(self, attr: &str) -> Self {
+            self.append_bytes(attr.as_bytes()).D8(0)
+        }
+
+        fn attr_ref1(self, attr: u8) -> Self {
+            self.D8(attr)
+        }
+
+        fn abbrev(self, code: u64, tag: DwTag, children: DwChildren) -> Self {
+            self.uleb(code).uleb(tag.0).D8(children.0)
+        }
+
+        fn abbrev_null(self) -> Self {
+            self.D8(0)
+        }
+
+        fn abbrev_attr(self, name: DwAt, form: DwForm) -> Self {
+            self.uleb(name.0).uleb(form.0)
+        }
+
+        fn abbrev_attr_null(self) -> Self {
+            self.D8(0).D8(0)
+        }
+
+        fn offset(self, offset: usize, format: Format) -> Self {
+            match format {
+                Format::Dwarf32 => self.L32(offset as u32),
+                Format::Dwarf64 => self.L64(offset as u64),
+            }
+        }
+
+        fn uleb(self, val: u64) -> Self {
+            let mut buf = Vec::new();
+            let written = leb128::write::unsigned(&mut buf, val).unwrap();
+            self.append_bytes(&buf[0..written])
+        }
```

Comment:
My `.eh_frame` branch will introduce `test_util.rs` which has a trait with `uleb` and `sleb` methods. Once I get that merged, we will be able to make this trait simply require that one, and we can remove the duplication.